### PR TITLE
feat: app.moveToApplicationsFolder conflict handling

### PR DESCRIPTION
**Pull request by codebytere**
#### Description of Change

Resolves https://github.com/electron/electron/issues/18805.

We want to keep default move conflict handling behavior in that it's still what most users would expect, but there exist edge cases in which users may not want to be forced into that behavior.

This thus introduces an optional conflict handler that allows developers access to more granular move actions. They could now allow the user to choose whether to delete an existing app in favor of the current one being moved, or whether to quit the current app and focus on the existing one should it both exist _and_ be running. I added a fair amount of new documentation outlining this behavior, but if there are things users may benefit from seeing examples of or nuances that should be added please leave feedback!

cc @MarshallOfSound @deepak1556 

#### Checklist

- [x] PR description included and stakeholders cc'd
- [x] `npm test` passes
- [x] relevant documentation is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/electron/electron/blob/master/docs/development/pull-requests.md#commit-message-guidelines)
- [x] [PR release notes](https://github.com/electron/clerk/blob/master/README.md) describe the change in a way relevant to app developers, and are [capitalized, punctuated, and past tense](https://github.com/electron/clerk/blob/master/README.md#examples).

#### Release Notes

Notes: Added an optional conflict handling callback to `app.moveToApplicationsFolder`.

### CODE REVIEW

File: docs/api/app.md [hunk 1]
```diff
@@ -1235,7 +1238,26 @@ the user to confirm the operation, you may do so using the
...
+you exactly what went wrong.
+
+By default, if an app of the same name as the one being moved exists in the Applications directory and is _not_ running, the existing app will be trashed and the active app moved into its place. If it _is_ running, the pre-existing running app will assume focus and the the previously active app will quit itself. This behavior can be changed by invoking the optional callback handler, where the boolean returned by the handler determines whether or not the move proceeds with its default behavior.
+
+For example:
+
+```js
+app.moveToApplicationsFolder(conflictType => {
+  if (conflictType === 'exists') {
+    dialog.showMessageBox({
```
Comment (MarshallOfSound) [inline docs/api/app.md]: This has to be sync right?

File: docs/api/app.md [hunk 2]
```diff
@@ -1222,7 +1222,10 @@ This method can only be called before app is ready.
 Returns `Boolean` - Whether the application is currently running from the
 systems Application folder. Use in combination with `app.moveToApplicationsFolder()`
 
-### `app.moveToApplicationsFolder()` _macOS_
+### `app.moveToApplicationsFolder([handler])` _macOS_
+
+* `handler` Function (optional) - A handler for potential conflict in move failure.
```
Comment (MarshallOfSound) [inline docs/api/app.md]: To allow cleaner API expansion this should probably be an `options` object with a single property `conflictHandler`

File: shell/browser/ui/cocoa/atom_bundle_mover.mm [hunk 3]
```diff
@@ -70,11 +90,22 @@
...
+    base::OnceCallback<bool(ConflictType)> conflict_cb;
+
     // If a copy already exists in the Applications folder, put it in the Trash
     if ([fileManager fileExistsAtPath:destinationPath]) {
       // But first, make sure that it's not running
       if (IsApplicationAtPathRunning(destinationPath)) {
-        // Give the running app focus and terminate myself
+        // Check for callback handler and get user choice for open/quit
+        if (args->GetNext(&conflict_cb)) {
+          return false;
```
Comment (MarshallOfSound) [inline shell/browser/ui/cocoa/atom_bundle_mover.mm]: I think this got left here by accident

File: shell/browser/ui/cocoa/atom_bundle_mover.mm [hunk 4]
```diff
@@ -70,11 +90,22 @@
       }
     }
   } else {
+    base::OnceCallback<bool(ConflictType)> conflict_cb;
```
Comment (MarshallOfSound) [inline shell/browser/ui/cocoa/atom_bundle_mover.mm]: What happens if this method returns something that isn't a `bool` (like a string or something).  I have a terrible feeling it's gonna throw a weird mate error or something

Comment (codebytere) [reply shell/browser/ui/cocoa/atom_bundle_mover.mm]: yeah, i think this is gonna need to be 
```
base::OnceCallback<v8::MaybeLocal<v8::Boolean>(ConflictType)> conflict_cb;
```

File: docs/api/app.md [hunk 5]
```diff
@@ -1222,7 +1222,11 @@ This method can only be called before app is ready.
 Returns `Boolean` - Whether the application is currently running from the
 systems Application folder. Use in combination with `app.moveToApplicationsFolder()`
 
-### `app.moveToApplicationsFolder()` _macOS_
+### `app.moveToApplicationsFolder([options])` _macOS_
+
+* `options` Object (optional)
+  * `conflictHandler` Function (optional) - A handler for potential conflict in move failure.
```
Comment (MarshallOfSound) [inline docs/api/app.md]: We can declare the return type 👍 

```suggestion
  * `conflictHandler` Function<Boolean> (optional) - A handler for potential conflict in move failure.
```

File: docs/api/app.md [hunk 6]
```diff
@@ -1235,7 +1239,28 @@ the user to confirm the operation, you may do so using the
 move to fail. For instance if the user cancels the authorization dialog, this
 method returns false. If we fail to perform the copy, then this method will
 throw an error. The message in the error should be informative and tell
-you exactly what went wrong
+you exactly what went wrong.
+
+By default, if an app of the same name as the one being moved exists in the Applications directory and is _not_ running, the existing app will be trashed and the active app moved into its place. If it _is_ running, the pre-existing running app will assume focus and the the previously active app will quit itself. This behavior can be changed by invoking the optional callback handler, where the boolean returned by the handler determines whether or not the move conflict is resolved with default behavior.
```
Comment (MarshallOfSound) [inline docs/api/app.md]:
```suggestion
By default, if an app of the same name as the one being moved exists in the Applications directory and is _not_ running, the existing app will be trashed and the active app moved into its place. If it _is_ running, the pre-existing running app will assume focus and the the previously active app will quit itself. This behavior can be changed by providing the optional conflict handler, where the boolean returned by the handler determines whether or not the move conflict is resolved with default behavior.  i.e. returning `false` will ensure no further action is taken, returning `true` will result in the default behavior and the method continuing. 
```

File: docs/api/app.md [hunk 7]
```diff
@@ -1235,7 +1239,28 @@ the user to confirm the operation, you may do so using the
...
+
+For example:
+
+```js
+app.moveToApplicationsFolder({
+  conflictHandler: conflictType => {
+    if (conflictType === 'exists') {
+      dialog.showMessageBoxSync({
```
Comment (MarshallOfSound) [inline docs/api/app.md]: This won't work if I'm reading this right, need to do `return dialog.showMessageBoxSync(...) === 1`, sync doesn't have a callback

File: shell/browser/ui/cocoa/atom_bundle_mover.h [hunk 8]
```diff
@@ -11,6 +11,9 @@
 
 namespace electron {
 
+// Possible bundle movement conflicts
+enum class ConflictType { EXISTS, EXISTS_AND_RUNNING };
```
Comment (MarshallOfSound) [inline shell/browser/ui/cocoa/atom_bundle_mover.h]: Just to make this slight more scoped than "an electron conflict"

```suggestion
enum class BundlerMoverConflictType { EXISTS, EXISTS_AND_RUNNING };
```

File: docs/api/app.md (hunk 7 above)
Comment (codebytere) [reply docs/api/app.md]: oh derp yes

**Comment by codebytere** [comment]
@MarshallOfSound it throws this error:
for both 1) non-boolean return types and 2) manually thrown Errors 

Screen Shot 2019-07-01 at 9 29 00 PM

**Comment by codebytere** [comment]
@MarshallOfSound this should be all set for final review!

**Comment by fabiospampinato** [comment]
@codebytere I feel this PR completely solves a specific issue: being able to show a dialog in case a conflict arises.

But if an app is already present _and_ running I guess what most people would want to do is quit the old app and then move the new one in, this PR at best would allow me to show the user an error, but what if I just want to quit the old app and proceed? Which would be the most common scenario under this situation IMHO. Is/shoudn't there be an easy way to do that?

And shouldn't an asynchronous `conflictHandler` function be supported too? Say I want to kill the other app via the shell and I want to wait for it to be actually killed before proceeding.

Also if I don't provide a `conflictHandler` is this method still returning `true` in conflicting situations? That would go against the docs that say that an error should be thrown instead: "If we fail to perform the copy, then this method will throw an error."

**Comment by codebytere** [comment]
@fabiospampinato this PR was created to address some use cases; it's unlikely we ever be able to cover all of them.

> I don't provide a conflictHandler is this method still returning true in conflicting situations? 
No, it does not change previous behavior. All new behavior is _optional_ as documented.

**Comment by fabiospampinato** [comment]
> No, it does not change previous behavior. 

The previous behavior is to return `true`, so shouldn't the answer here be "yes" and either the behavior or the docs updated as the two are conflicting? The app is not moved and no error is thrown, which was point of my issue that has been closed because of this PR.

**Pull request merged**